staking/README.md

```diff
@@ -0,0 +1,541 @@
+
+The `PostHandleMsgValidatorExit` post-handler then sets the deactivation epoch for the validator and persists in the state via the keeper:
+
+```
+validator, ok := k.GetValidatorFromValID(ctx, msg.ID)
+if !ok {
+	k.Logger(ctx).Error("Fetching of validator from store failed", "validatorID", msg.ID)
+	return hmCommon.ErrNoValidator(k.Codespace()).Result()
+}
+
+// set end epoch
+validator.EndEpoch = msg.DeactivationEpoch
+
+// update last updated
+validator.LastUpdated = sequence.String()
+
+// update nonce
+validator.Nonce = msg.Nonce
+
+// Add deactivation time for validator
+if err := k.AddValidator(ctx, validator); err != nil {
+	k.Logger(ctx).Error("Error while setting deactivation epoch to validator", "error", err, "validatorID", validator.ID.String())
+	return hmCommon.ErrValidatorNotDeactivated(k.Codespace()).Result()
+}
+```
+
+The `EndBlocker` hook then updates the validator set once deactivation epoch is completed.
+
+## How to propose a MsgValidatorExit transaction
+
+The `bridge` service in an existing validator's heimdall process polls for `UnstakeInit` event periodically and generates and broadcasts the transaction once it detects and parses the event. An existing validator on the network can also leverage the CLI to send the transaction:
+
+```
+heimdallcli tx staking validator-exit --proposer <PROPOSER_ADDRESS> --id <VALIDATOR_ID> --tx-hash <ETH_TX_HASH> --nonce <VALIDATOR_NONCE> --log-index <LOG_INDEX> --block-number <BLOCK_NUMBER> --deactivation-epoch <DEACTIVATION_EPOCH>
+```
+
+Or the REST server :
+
+```
+curl -X POST "localhost:1317/staking/validator-exit?from=<PROPOSER_ADDRESS>&id=<VALIDATOR_ID>&tx-hash=<ETH_TX_HASH>&nonce=<VALIDATOR_NONCE>&log-index=<LOG_INDEX>&block-number=<BLOCK_NUMBER>&deactivation-epoch=<DEACTIVATION_EPOCH>"
+```
+
+## How does a validator update its stake
+
+A validator can update its stake in the network by invoking the `restake` function on the `StakeManager` contract on L1, which emits an `StakeUpdate` event:
+
+```
+/// @param validatorId unique integer to identify a validator.
+/// @param nonce to synchronize the events in heimdall.
+/// @param newAmount the updated stake amount.
+event StakeUpdate(
+    uint256 indexed validatorId,
+    uint256 indexed nonce,
+    uint256 indexed newAmount
+);
+```
+
+On Heimdall, this event is parsed and a `MsgStakeUpdate` transaction is broadcasted, which is represented by the data structure:
+
+```
+type MsgStakeUpdate struct {
+	From        hmTypes.HeimdallAddress `json:"from"`
+	ID          hmTypes.ValidatorID     `json:"id"`
+	NewAmount   sdk.Int                 `json:"amount"`
+	TxHash      hmTypes.HeimdallHash    `json:"tx_hash"`
+	LogIndex    uint64                  `json:"log_index"`
+	BlockNumber uint64                  `json:"block_number"`
+	Nonce       uint64                  `json:"nonce"`
+}
+```
+where ,
```

Comment:
i'd remove the comma

Reply:
Done: [159a33b](https://github.com/maticnetwork/heimdall/pull/1082/commits/159a33bafa2f43c7cc2ae0f9d57b4257ef036436)